library/network/src/lib/network/firewalld.rb

```diff
@@ -0,0 +1,324 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+#
+# ***************************************************************************
+#
+# File: lib/network/firewalld.rb
+# Summary:  FirewallD configuration API
+# Authors:  Karol Mroz <kmroz@suse.de>, Markos Chandras <mchandras@suse.de>
+#
+
+require 'yast'
+
+module Firewalld
+
+  # exception when firewall-cmd fails
+  class FirewallCMDError < StandardError
+  end
+
+  class FWCmd
+
+    include Yast::Logger
+
+    BASH_SCR_PATH = Yast::Path.new(".target.bash_output")
+    # Base firewall-cmd command
+    COMMAND = "LANG=Python firewall-cmd"
+
+    attr_reader :option_str
+
+    def initialize(option_str)
+      @option_str = option_str
+    end
+
+    def command
+      "#{COMMAND} #{option_str}".strip.squeeze(" ")
+    end
+
+    # Output resulting of executing the command
+    def output(need_output = true)
+      cmd_result = Yast::SCR.Execute(BASH_SCR_PATH, command)
+
+      # See firewall-cmd manpage for exit codes. Not all of them justify an
+      # exception. 0 and 1 can be used as true and false respectively.
+      case cmd_result["exit"]
+      when 0, 1
+        if need_output
+          cmd_result["stdout"]
+        else
+          log.debug "#{command} returned: #{cmd_result['stdout']}"
+          cmd_result["exit"].zero? ? true : false
+        end
+      else
+        raise FirewallCMDError, "Calling firewall-cmd (cmd: #{command}) failed: #{cmd_result["stderr"]}"
+      end
+    end
+  end
+
+  class FirewalldAPI
+    def self.create(type = :bash)
+      case type
+      when :bash
+        FirewalldBashAPI.new
+      when :dbus
+        nil
+      else
+        raise "Unsupported Firewalld API type: #{type}"
+      end
+    end
+  end
+
+  class FirewalldBashAPI
+
+    # Simple wrapper for commands. Returns true on success
+    def do_cmd(*args)
+      fwcmd = FWCmd.new(args.join(''))
+      fwcmd.output(false)
+    end
+
+    ### State ###
+
+    # True if firewalld is running
+    def is_running?
+      do_cmd("--state")
+    end
+
+    # True if firewalld was reloaded
+    def reload
+      do_cmd("--reload")
+    end
+
+    # True if firewalld was fully reloaded
+    def complete_reload
+      do_cmd("--complete-reload")
+    end
+
+    # True if runtime config has been made permanent
+    def make_permanent
+      do_cmd("--runtime-to-permanent")
+    end
+
+    ### Zones ####
+
+    # Return list of zones.
+    def get_zones
+      fwcmd = FWCmd.new("--permanent --get-zones")
+      fwcmd.output.split(" ")
+    end
+
+    # Return zone, or nil.
```

Comment:
No. `do_cmd` returns `true` or `false`.
BTW "do_cmd" and "output" are poor names with respect to what they return.